src/main/java/dev/blaauwendraad/masker/json/ByteTrie.java

```diff
@@ -84,34 +91,60 @@ public void insert(String word) {
                         1. Locale issues (see String#equalsIgnoreCase)
                         2. So we don't have to convert when searching
                      */
-                    node.children[lowerBytes[i] + MAX_BYTE_SIZE] = child;
-                    node.children[upperBytes[i] + MAX_BYTE_SIZE] = child;
+                    node.children[lowerBytes[i] + BYTE_OFFSET] = child;
+                    node.children[upperBytes[i] + BYTE_OFFSET] = child;
                 }
             }
             node = child;
         }
+        node.keyMaskingConfig = config.getConfig(word);
         node.endOfWord = true;
+        node.negativeMatch = negativeMatch;
     }
 
     /**
-     * Returns if the word is in the trie.
+     * If the key matches - returns its masking configuration, otherwise returns null.
      */
     public boolean search(byte[] bytes, int offset, int length) {
+        TrieNode node = searchNode(bytes, offset, length);
+        return node != null && !node.negativeMatch;
+    }
+
+    /**
+     * Must be called only if the key has matched or in allow mode, for key that did not match
+     * returns its masking configuration, otherwise returns default config.
+     */
+    public KeyMaskingConfig config(byte[] bytes, int offset, int length) {
```

Comment:
Let me get back to this some time after, I'm still not sure whether I want to do separate `search`/`config`. Ideally it should be one method

Reply:
I am in favor of having one method that returns a TrieNode. Then the client can decide wether it wants to access a config or check for the match. This will also remove the binding of the allow mode logic to the trie.

Reply:
@donavdey I've solved it by returning a config instead, now that `ByteTrie` does much more than just a trie, I have renamed it to `KeyMatcher`